cmd/ingester/app/consumer/consumer.go

```diff
@@ -29,153 +28,116 @@ import (
+	internalConsumer    consumer.Consumer
+	processorFactory    ProcessorFactory
+	partitionsHeld      int64
+	partitionsHeldGauge metrics.Gauge
+}
 
-	internalConsumer consumer.Consumer
-	processorFactory ProcessorFactory
+type consumerGroupHandler struct {
+	processorFactory         ProcessorFactory
+	partitionToProcessor     map[int32]processor.SpanProcessor
+	logger                   *zap.Logger
+	consumer                 *Consumer
+	partitionToProcessorLock sync.RWMutex
+}
 
-	deadlockDetector deadlockDetector
+func (h *consumerGroupHandler) Setup(session sarama.ConsumerGroupSession) error {
+	h.partitionToProcessor = map[int32]processor.SpanProcessor{}
+	return nil
+}
 
-	partitionIDToState  map[int32]*consumerState
-	partitionMapLock    sync.Mutex
-	partitionsHeld      int64
-	partitionsHeldGauge metrics.Gauge
+func (h *consumerGroupHandler) getProcessFactory(session sarama.ConsumerGroupSession, partition int32, offset int64) processor.SpanProcessor {
+	h.partitionToProcessorLock.RLock()
+	msgProcessor := h.partitionToProcessor[partition]
+	h.partitionToProcessorLock.RUnlock()
+	if msgProcessor == nil {
+		msgProcessor = h.processorFactory.new(session, partition, offset-1)
+		h.partitionToProcessorLock.Lock()
+		h.partitionToProcessor[partition] = msgProcessor
+		h.partitionToProcessorLock.Unlock()
+	}
+	return msgProcessor
 }
+func (h *consumerGroupHandler) Cleanup(_ sarama.ConsumerGroupSession) error { return nil }
+func (h *consumerGroupHandler) ConsumeClaim(sess sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	msgMetrics := h.consumer.newMsgMetrics(claim.Partition())
 
-type consumerState struct {
-	wg                sync.WaitGroup
-	partitionConsumer sc.PartitionConsumer
+	for {
+		select {
+		case msg, ok := <-claim.Messages():
+			if !ok {
+				h.logger.Info("Message channel closed. ", zap.Int32("partition", claim.Partition()))
+				return nil
+			}
+			h.logger.Debug("Got msg", zap.Any("msg", msg))
+			msgMetrics.counter.Inc(1)
+			msgMetrics.offsetGauge.Update(msg.Offset)
+			msgMetrics.lagGauge.Update(claim.HighWaterMarkOffset() - msg.Offset - 1)
+			msgProcessor := h.getProcessFactory(sess, claim.Partition(), msg.Offset)
+			msgProcessor.Process(&saramaMessageWrapper{msg})
+		}
+	}
+	return nil
 }
 
 // New is a constructor for a Consumer
 func New(params Params) (*Consumer, error) {
-	deadlockDetector := newDeadlockDetector(params.MetricsFactory, params.Logger, params.DeadlockCheckInterval)
 	return &Consumer{
 		metricsFactory:      params.MetricsFactory,
 		logger:              params.Logger,
 		internalConsumer:    params.InternalConsumer,
 		processorFactory:    params.ProcessorFactory,
-		deadlockDetector:    deadlockDetector,
-		partitionIDToState:  make(map[int32]*consumerState),
 		partitionsHeldGauge: partitionsHeldGauge(params.MetricsFactory),
 	}, nil
 }
 
 // Start begins consuming messages in a go routine
 func (c *Consumer) Start() {
-	c.deadlockDetector.start()
```

Comment:
I know this was added after the first implementation because of a specific problem that Uber experienced, although I don't remember what the problem was exactly. Are you aware of that? Is it safe to remove this?

Reply:
I'm aware of that, and still not sure if this should be removed, Is something I'm investigating.. it seems like the problem was solved in this implementation. but still not 100% sure.

Reply:
Maybe deprecate as Juca suggests, but only start the deadlock detector if internal > 0 (I think 0 is the default and means no deadlock detection used)?

Reply:
It seems like this is not necesary anymore, the issue was fixed on sarama. https://github.com/Shopify/sarama/pull/1156

Anyway, I agree with the idea of preserve this deadlockDetector, just in case, and mark it as deprecated. I'll reintroduce it.